src/java/org/apache/cassandra/db/Keyspace.java

```diff
@@ -148,19 +157,19 @@ public static Keyspace clear(String keyspaceName)
         return clear(keyspaceName, Schema.instance);
     }
 
+    /**
+     * Note that this method should only be called while holding the monitor lock for {@link Schema#instance} 
+     */
     public static Keyspace clear(String keyspaceName, Schema schema)
     {
-        synchronized (Keyspace.class)
+        Keyspace t = schema.removeKeyspaceInstance(keyspaceName);
```

Comment:
should lock on `schema` as well here to match.

Reply:
@dcapwell All codepaths that arrive here should already be holding the `Schema.instance` monitor lock, but reentrancy should make it harmless. It's probably better, in retrospect, than a method comment.